python/rmm/rmm.py

```diff
@@ -130,107 +130,72 @@ def csv_log():
+    details of the interface being implemented here.
     """
-    shape, strides, dtype = cuda.api._prepare_shape_strides_dtype(
-        shape, strides, dtype, order
-    )
-    datasize = cuda.driver.memory_size_from_info(
-        shape, strides, dtype.itemsize
-    )
 
-    buf = librmm.DeviceBuffer(size=datasize, stream=stream)
-
-    ctx = cuda.current_context()
-    ptr = ctypes.c_uint64(int(buf.ptr))
-    mem = cuda.driver.MemoryPointer(ctx, ptr, datasize, owner=buf)
-    return cuda.cudadrv.devicearray.DeviceNDArray(
-        shape, strides, dtype, gpu_data=mem
-    )
-
-
-def device_array_like(ary, stream=0):
-    """
-    device_array_like(ary, stream=0)
-
-    Call rmmlib.device_array with information from `ary`. Clone of Numba
-    `cuda.device_array_like`, but uses RMM for device memory management.
-    """
-    if ary.ndim == 0:
-        ary = ary.reshape(1)
+    def initialize(self):
+        # No special initialization needed to use RMM within a given context.
+        pass
 
-    return device_array(ary.shape, ary.dtype, ary.strides, stream=stream)
+    def memalloc(self, size):
+        """
+        Allocate an on-device array from the RMM pool.
+        """
+        buf = librmm.DeviceBuffer(size=size)
+        ctx = self.context
+        ptr = ctypes.c_uint64(int(buf.ptr))
+        finalizer = _make_finalizer(ptr.value, self.allocations)
 
+        # self.allocations is initialized by the parent, HostOnlyCUDAManager,
+        # and cleared upon context reset, so although we insert into it here
+        # and delete from it in the finalizer, we need not do any other
+        # housekeeping elsewhere.
+        self.allocations[ptr.value] = buf
 
-def to_device(ary, stream=0, copy=True, to=None):
-    """
-    to_device(ary, stream=0, copy=True, to=None)
+        return MemoryPointer(ctx, ptr, size, finalizer=finalizer)
 
-    Allocate and transfer a numpy ndarray or structured scalar to the device.
-    Clone of Numba `cuda.to_device`, but uses RMM for device memory management.
-    """
-    if to is None:
-        to = device_array_like(ary, stream=stream)
-        to.copy_to_device(ary, stream=stream)
-        return to
-    if copy:
-        to.copy_to_device(ary, stream=stream)
-    return to
+    def get_ipc_handle(self, memory):
+        """
+        Get an IPC handle for the MemoryPointer memory with offset modified by
+        the RMM memory pool.
+        """
+        ipchandle = (ctypes.c_byte * 64)()  # IPC handle is 64 bytes
+        cuda.cudadrv.driver.driver.cuIpcGetMemHandle(
+            ctypes.byref(ipchandle), memory.owner.handle,
+        )
+        source_info = cuda.current_context().device.get_device_identity()
+        ptr = memory.device_ctypes_pointer.value
+        offset = librmm.rmm_getallocationoffset(ptr, 0)
+        return IpcHandle(
+            memory, ipchandle, memory.size, source_info, offset=offset
+        )
 
+    def get_memory_info(self):
```

Comment:
Just curious if this is essential functionality or a nice-to-have?

I'm working on removing the use of deprecated RMM APIs in https://github.com/rapidsai/rmm/pull/361, `rmmGetInfo()` being one of those.

Further, the new `memory_resource` classes don't all support querying memory information, so depending on the memory resource being used, we may or may not be able to offer this functionality.

cc: @jrhemstad

Reply:
Just a nice-to-have - it's allowed to raise `NotImplementedError`: http://numba.pydata.org/numba-doc/latest/cuda/external-memory.html#numba.cuda.BaseCUDAMemoryManager.get_memory_info

Reply:
Thanks!